### DISCUSSION TITLE
Define maybeparse(T,x)::Union{Nothing,Some{T}}

### DISCUSSION THREAD

**Discussion by adkabo** [JuliaLang/julia#43746] [General]
### Background

A number of functions in Base and packages return `T` or `nothing`, often written before Julia had `Some{T}`. Callers of these functions are not always able to tell whether the computation failed or successfully returned a value of `nothing`. 

### Proposal

I want to use this pattern:

```jl
y = maybeparse(T,x)::Union{Nothing,Some{T}}
return isnothing(y) ? nothing : handle(something(y))
```

`tryparse` doesn't wrap the successful value in `Some`. `maybeparse` would do so and would be a replacement for that function to be recommended in new code.

This function isn't strictly necessary since we already can distinguish using `try parse(T,x) catch` but it would be nice to have an exceptionless version that can be used in expression contexts like [pattern matching](https://thautwarm.github.io/MLStyle.jl/latest/syntax/pattern.html). 

### Naming

`maybeparse` is a bit long so it could be `mayparse`.

### References

- [A blog post](https://codewords.recurse.com/issues/one/option-and-null-in-dynamic-languages) about why this wrapping is good
- A similar issue for `get` https://github.com/JuliaLang/julia/pull/34821
- A similar issue for `maybetake!` https://github.com/JuliaLang/julia/pull/41966 by @tkf 
- A similar issue for `findfirst` https://github.com/JuliaLang/julia/issues/29565 (Mentioned in comment below.)
- A `tryparse` issue https://github.com/JuliaLang/julia/issues/43149

**Comment by nalimilan** [comment]
An alternative solution which would fit better with the current API would be to write `tryparse(Some{T}, x)` instead of `maybeparse(T, x)`. I think we should make a general design choice regarding whether we want to add `maybe*` functions or use a different pattern.

EDIT: see also https://github.com/JuliaLang/julia/issues/29565 where we need a syntax to call `findfirst` and get a `Union{Some, Nothing}` result.

**Comment by adkabo** [comment]
I don't like the idea of adding syntactic overhead (`Some{T}`) to buy correctness, so my preference is to move away from the current API rather than trying to stick with it. The downside is growing API surface, the impact of which can be addressed here by soft-deprecating the old version in the docs.

**Comment by tkf** [comment]
I've been wondering about something similar but a bit generalized version: Can we have a consistent convention/API for `Union{Some{T},Nothing}`-valued and "result/either type"-valued [^1] functions?

We have the `f!` naming scheme for destructive functions and it has been working quite nicely in practice. Likewise, it's nice to have a standard way to tell that a given function uses a certain convention or API. It's perhaps more true for maybe-valued and result-valued functions than `f!` because we can have a standard toolset for composing such functions like Rust's `and_then` and `or_else`.

[^1]: By result type I'm talking about something like `Union{Ok{T},Err{E}}` with `struct Ok{T}; value::T; end` and `struct Err{E<:Exception}; value::E; end`. There's a latest summary and discussion by @jakobnissen in [[ANN] ErrorTypes.jl - Rust-like safe errors in Julia - Package Announcements / Package announcements - JuliaLang](https://discourse.julialang.org/t/ann-errortypes-jl-rust-like-safe-errors-in-julia/53953) and also my exploration around this topic https://github.com/tkf/Try.jl.

> `tryparse(Some{T}, x)` instead of `maybeparse(T, x)`

I don't think this can work *if* we want to make it possible to generalize `tryparse`-like API to non-`Number`s because `tryparse(Some{Int}, "Some(1)")` is ambiguous.

**Comment by jakobnissen** [comment]
First, let me say that I am completely, totally on board with having a concerted effort to create a well-thought out system for result types in Julia. These kinds of systems are great in any language, but Julia seems to be uniquely suitable for creating a nice result type system due to its type system.

The current situation with `Some` and `Nothing` seems designed _ad hoc_ with a whole bunch of edge cases, and is not exactly ideal. I don't think the `maybeparse` proposal is helpful in getting us to a more general, principled design. It would just be one more incomplete design.

With regard to a naming scheme for failable functions, I would prefer using type annotation on functions for this instead of a naming convention. Not only are type asserts easy to read and do not introduce new syntax or concepts, they actually _assert_ that the function returns a result type instead of just suggesting it. Further, they can optionally say exactly _which_ result type.

**Comment by vtjnash** [comment]
I moved this to a discussion thread, since it does not have a specific feature or issue request, but is a good discussion to have

**Comment by jakobnissen** [comment]
See also discussion here: https://julialang.zulipchat.com/#narrow/stream/137791-general/topic/Try.2Ejl

**Comment by tkf** [comment]
I don't want to hijack adkabo's thread on `Union{Nothing,Some{T}}`. So I opened #43773 that focuses on result values ("`Ok`/`Err`") where the error values can have more information than `nothing`.

**Reply by adkabo** [reply]
+1. Having Ok/Err in Base would be a much more satisfactory solution than my narrow proposal here.
